sonarqube/resource_sonarqube_project_main_branch.go

```diff
@@ -0,0 +1,139 @@
+	return &schema.Resource{
+		Create: resourceSonarqubeProjectMainBranchCreate,
+		Read:   resourceSonarqubeProjectMainBranchRead,
+		Delete: resourceSonarqubeProjectMainBranchDelete,
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceSonarqubeProjectMainBranchCreate(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/project_branches/rename"
+
+	sonarQubeURL.RawQuery = url.Values{
+		"name":    []string{d.Get("name").(string)},
+		"project": []string{d.Get("project").(string)},
+	}.Encode()
+
+	resp, err := httpRequestHelper(
+		m.(*ProviderConfiguration).httpClient,
+		"POST",
+		sonarQubeURL.String(),
+		http.StatusNoContent,
+		"resourceSonarqubeProjectMainBranchCreate",
+	)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	id := fmt.Sprintf("%v/%v", d.Get("project").(string), d.Get("name").(string))
+	d.SetId(id)
+
+	return resourceSonarqubeProjectMainBranchRead(d, m)
+}
+
+func resourceSonarqubeProjectMainBranchRead(d *schema.ResourceData, m interface{}) error {
+	idSlice := strings.Split(d.Id(), "/")
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/project_branches/list"
+	sonarQubeURL.RawQuery = url.Values{
+		"project": []string{idSlice[0]},
+	}.Encode()
+
+	resp, err := httpRequestHelper(
+		m.(*ProviderConfiguration).httpClient,
+		"GET",
+		sonarQubeURL.String(),
+		http.StatusOK,
+		"resourceSonarqubeProjectMainBranchRead",
+	)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Decode response into struct
+	branchReadResponse := GetBranches{}
+	err = json.NewDecoder(resp.Body).Decode(&branchReadResponse)
+	if err != nil {
+		return fmt.Errorf("resourceSonarqubeProjectMainBranchRead: Failed to decode json into struct: %+v", err)
+	}
+	// Loop over all branches to see if the main branch we need exists.
+	for _, value := range branchReadResponse.Branches {
+		if idSlice[1] == value.Name && value.IsMain {
+			d.Set("project", idSlice[0])
+			d.Set("name", value.Name)
+			return nil
+		}
+	}
+	return fmt.Errorf("resourceSonarqubeProjectMainBranchRead: Failed to find project main branch: %+v", d.Id())
+
+}
+
+// TODO make the delete function read the default branch name of the sonarQube instance instead of assuming
+func resourceSonarqubeProjectMainBranchDelete(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/project_branches/rename"
+	sonarQubeURL.RawQuery = url.Values{
+		"name":    []string{"main"},
```

Comment:
Please get the default branch name from `[]string{d.Get("name").(string)}` instead of hardcoding main,

Reply:
The idea of delete was to revert the branch back to what is default for the sonarqube instance. That is main unless the setting has been changed by an administrator. Perhaps we can fetch it via the api for settings because we know the setting name. But i dont know how that would work as i dont write too much terraform provider. if that makes sense

Reply:
ah ok, i misunderstood. Does sonarqubes default branch change depending on which git provider is being used? I think if the default is always main we can leave this as a TODO. Shouldn't be too much work to implement.

If its possible to fetch this config via https://next.sonarqube.com/sonarqube/web_api/api/settings/values then that would be awesome. In the provider you would just make another http call on deletion to fetch that config.

This example is for on creation but its a similar idea. https://github.com/jdamata/terraform-provider-sonarqube/pull/120/files#diff-5ad43f98cfd73869c45ce8eede0af3226ba13143149e948a519dc6c363ed2f28R42 . First we call the isLocal() function to get some information about a user account before creating it.

Reply:
> ah ok, i misunderstood. Does sonarqubes default branch change depending on which git provider is being used? I think if the default is always main we can leave this as a TODO. Shouldn't be too much work to implement.
> 

![image](https://user-images.githubusercontent.com/23239792/216664223-b9267296-f550-4753-891e-9aaad98fc914.png)
Sonarqube has a Rule that sets the default main branch name , this is by default main but it can be changed by the user. Be best solution would be if we can , when deleting the github binding that we just revert to the value of this setting. I hardcoded it main because i dont know how that logic would look like, or if its ok to call an api call that is actually unrelated to the deletetion itself if that makes sense. But because we already have functions in place to read rules from the settings provider then we could reuse that logic to fetch the rule. If you write a pseudo perhaps i can implement it, it shouldnt be too hard i think? But i think we can leave it as todo for now, i dont mind comming back to this functionality, as im probably also doing imports in a future pr. I just need this for work right now ^^ 


> If its possible to fetch this config via https://next.sonarqube.com/sonarqube/web_api/api/settings/values then that would be awesome. In the provider you would just make another http call on deletion to fetch that config.
> 

Its possible to fetch it as i state above. We could fetch it as you talk about below in like an init step perhaps so we only have to call it once if thats possible.

> This example is for on creation but its a similar idea. https://github.com/jdamata/terraform-provider-sonarqube/pull/120/files#diff-5ad43f98cfd73869c45ce8eede0af3226ba13143149e948a519dc6c363ed2f28R42 . First we call the isLocal() function to get some information about a user account before creating it.

I dont want to have to make the user provide me with the branch name they wanna revert back to. If you were talking about specifically askign the user to provide like a "default_branch_name" attribute field we can read

Reply:
If you think my logic is wrong do tell me, This is the first time im writing anything for terraform and i've just recently learned using it. Thats why i havent done the import functions yet because i havent used them peronsally yet in my projects , so i dont actually know fully what it does 😄

Reply:
> I hardcoded it main because i dont know how that logic would look like, or if its ok to call an api call that is actually unrelated to the deletetion itself if that makes sense.

Yep definitely ok to make the http call to fetch the setting. It is related to deletion since we need that data. 

Im good leaving it as a TODO for now

Reply:
Then lets agree on leaving it a TODO for now. I will revisit it when i have more time :)